### PR TITLE
Remove existing property - Part 6, IFluidDataStoreRuntime

### PR DESCRIPTION
**Pull request by andre4i**
See #3429 

- the `existing` property from `IFluidDataStoreRuntime` (and `FluidDataStoreRuntime`) has been removed.
- The `IFluidDataStoreFactory.instantiateDataStore` has an optional `existing` parameter which can be used to replace checking  `IFluidDataStoreContext.existing`.
- the `existing` property from `IFluidDataStoreContext` has been marked as deprecated and will be removed in +2 releases. Left in place a back-compat way of dealing with it.

### CODE REVIEW

File: examples/data-objects/key-value-cache/src/index.ts [hunk 1]
```diff
@@ -88,8 +88,8 @@ class KeyValue implements IKeyValue, IFluidObject, IFluidRouter {
         return defaultFluidObjectRequestHandler(this, request);
     }
 
-    private async initialize() {
-        if (!this.runtime.existing) {
+    private async initialize(existing: boolean) {
+        if (!existing) {
```
Comment (vladsud) [inline examples/data-objects/key-value-cache/src/index.ts:92]: I'm not sure it's the right time to consider it, but worth sharing:
I'm looking more and more at creation of things as declaration of schema. I.e. if we look at fluid objects as one big graph, types of DDSs / Data stores declare the meaning / interpretation of data (and ops) in some sub-sections of this graph. Creation of DDS is essentially a declaration of schema for that part of container graph.
And logical continuation here - concurrent non-conflicting schema declaration for same node should be allowed and it should coalesce. I.e. it should be fine to create a map with this name if it is already there - it should be a noop.
This direction works perfectly if initialization is empty, i.e. like here - we do not put any content into the map on creation (otherwise initialization can overwrite user content that another client put into it).

So maybe most of our examples should not even use existing flag, and instead we should always "create" map, and make sure it does not explode on collision (probably needs to be opt-in behavior)

File: examples/data-objects/shared-text/src/document.ts [hunk 2]
```diff
@@ -13,33 +13,27 @@ const rootMapId = "root";
...
 
     /**
      * Constructs a new document from the provided details
      */
-    private constructor(public runtime: IFluidDataStoreRuntime, private readonly root: ISharedMap) {
-    }
+     private constructor(
+        public runtime: IFluidDataStoreRuntime,
+        private readonly root: ISharedMap,
+        public readonly existing: boolean,
```
Comment (vladsud) [inline examples/data-objects/shared-text/src/document.ts:35]: Is there a way to not have existing on this class?
It's ok as is (as we push it out from runtime into these leafy cases), but it would be even better if it was gone here as well

File: packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts [hunk 3]
```diff
@@ -60,14 +62,15 @@ async function createDataObject<TObj extends PureDataObject<O, S, E>, O, S, E ex
     // request mixin in
     runtimeClass = mixinRequestHandler(
         async (request: IRequest, runtimeArg: FluidDataStoreRuntime) =>
-            (await PureDataObject.getDataObject(runtimeArg)).request(request),
+            (await PureDataObject.getDataObject(runtimeArg, existing)).request(request),
```
Comment (vladsud) [inline packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts:65]: I think we should avoid adding existing here.
Basically code below (end of this function) ensures that initialization is fully complete for create flow.
So any call to getDataObject() is done to fully initialized object, and thus existing it passes to finishInitialization should be always false.
getDataObject() is used in Bohemia (I think), so I'd try to keep signature the same.

File: packages/framework/aqueduct/src/data-objects/dataObject.ts [hunk 4]
```diff
@@ -62,8 +62,10 @@ export abstract class DataObject<O extends IFluidObject = object, S = undefined,
      * Initializes internal objects and calls initialization overrides.
      * Caller is responsible for ensuring this is only invoked once.
      */
-    public async initializeInternal(): Promise<void> {
-        if (!this.runtime.existing) {
+    public async initializeInternal(existing?: boolean): Promise<void> {
+        const backCompatExisting = instantiateExisting(this.context, existing);
```
Comment (vladsud) [inline packages/framework/aqueduct/src/data-objects/dataObject.ts:66]: I'm not sure this is required. this method is called only by finishInitialization() that passes boolean (no undefined). Tracing code up, you already do instantiateExisting() dance in instantiateDataStore(), so all the code down operates with boolean

File: packages/runtime/client-api/src/api/document.ts [hunk 5]
```diff
@@ -90,6 +83,7 @@ export class Document extends EventEmitter {
         public readonly context: IFluidDataStoreContext,
         private readonly root: ISharedMap,
         private readonly closeFn: () => void,
+        public readonly existing: boolean,
```
Comment (vladsud) [inline packages/runtime/client-api/src/api/document.ts:86]: Can we remove it? :)

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 6]
```diff
@@ -217,7 +217,7 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
         // Thus having slashes in types almost guarantees trouble down the road!
         assert(id.indexOf("/") === -1, 0x13a /* `Data store ID contains slash: ${id}` */);
 
-        this._attachState = this.containerRuntime.attachState !== AttachState.Detached && existing ?
+        this._attachState = this.containerRuntime.attachState !== AttachState.Detached && this.existing ?
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:220]: can this.existing be private?

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 7]
```diff
@@ -945,7 +944,7 @@ export class LocalDetachedFluidDataStoreContext
             scope,
             createSummarizerNode,
             bindChannel,
-            snapshotTree,
+            undefined,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:947]: Is this related, or incorrect merge conflict? I'm not sure I see relationship to your changes.

**Review by vladsud — APPROVED** [review]

File: packages/runtime/container-runtime/src/dataStoreContext.ts (hunk 7 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:947]: `LocalDetachedFluidDataStoreContext`'s snapshotTree is always `undefined` (ending up being non-'existent') so I just thought to make this more explicit..

File: packages/runtime/runtime-utils/src/utils.ts [hunk 8]
```diff
@@ -26,3 +27,15 @@ export type ReadAndParseBlob = <T>(id: string) => Promise<T>;
...
 }
+
+/**
+ * @deprecated #6346
+ *
+ * Backwards compatible way of deciding whether we're instantiating new or from an existing
+ * context. To be removed after `existing` is a required parameter in
+ * IFluidDataStoreFactory.instantiateDataStore
+ */
+export const instantiateExisting = (context: IFluidDataStoreContext, existing?: boolean): boolean =>
```
[the remark this answers is not in the document]
Comment (andre4i) [reply packages/runtime/runtime-utils/src/utils.ts:38]: naming is hard. Will redo

File: examples/data-objects/shared-text/src/document.ts (hunk 2 above)
Comment (andre4i) [reply examples/data-objects/shared-text/src/document.ts:35]: There are some references to it. Let me see if I can do something about it..

File: packages/framework/aqueduct/src/data-objects/dataObject.ts (hunk 4 above)
Comment (andre4i) [reply packages/framework/aqueduct/src/data-objects/dataObject.ts:66]: This is the end of the (long) story of my back-compat testing. I'll check to see if I can simplify it. Thanks!

File: packages/runtime/client-api/src/api/document.ts (hunk 5 above)
Comment (andre4i) [reply packages/runtime/client-api/src/api/document.ts:86]: There are some references to it unfortunately. I'll check if I can refactor them.

File: packages/runtime/container-runtime/src/dataStoreContext.ts (hunk 6 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:220]: Not until we make the existing parameter mandatory inside `IFluidDataStoreFactory.instantiateDataStore`..

File: packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts (hunk 3 above)
Comment (andre4i) [reply packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts:65]: I can't find an elegant way of keeping this the same. I think we need to live with breaking this API.

The flow at the end of this function does indeed call this with `false`. However removing this in the mixin, will lead to breaking all other `existing` scenarios. I had clear manual reproes (outside of unit tests) with the clicker (for example) when the request processing with getdataobject would fail because the thing already exists (see `0x179`).

Comment (vladsud) [reply packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts:65]: I want to understand it better - let's chat.
Take a look at createOdspSearchDataStoreFactory in Bohemia - there is nowhere to get existing flag in that flow, unless we baske it into mixins (mixinRequestHandler, mixinSummaryHandler), i.e. existing is provided by mixin into mixin callback, not captured from ambient environment (createDataObject() closure).

Comment (andre4i) [reply packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts]: synced offline. Thanks, Vlad!
